// movier/dist/src/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMDBPathType = exports.ImageType = exports.AwardOutcome = exports.Language = exports.TitleMainType = exports.Genre = exports.Source = void 0;
var Source;
(function (Source) {
    Source["IMDB"] = "IMDB";
    Source["MetaCritics"] = "metaCritics";
    Source["Unknown"] = "unknown";
})(Source = exports.Source || (exports.Source = {}));
var Genre;
(function (Genre) {
    Genre["Action"] = "action";
    Genre["Adventure"] = "adventure";
    Genre["Animation"] = "animation";
    Genre["Biography"] = "biography";
    Genre["Comedy"] = "comedy";
    Genre["Crime"] = "crime";
    Genre["Documentary"] = "documentary";
    Genre["Drama"] = "drama";
    Genre["Family"] = "family";
    Genre["Fantasy"] = "fantasy";
    Genre["FilmNoir"] = "filmNoir";
    Genre["GameShow"] = "gameShow";
    Genre["History"] = "history";
    Genre["Horror"] = "horror";
    Genre["Music"] = "music";
    Genre["Musical"] = "musical";
    Genre["Mystery"] = "mystery";
    Genre["News"] = "news";
    Genre["RealityTV"] = "realityTv";
    Genre["Romance"] = "romance";
    Genre["SciFi"] = "sciFi";
    Genre["Short"] = "short";
    Genre["Sport"] = "sport";
    Genre["TalkShow"] = "talkShow";
    Genre["Thriller"] = "thriller";
    Genre["War"] = "war";
    Genre["Western"] = "western";
})(Genre = exports.Genre || (exports.Genre = {}));
var TitleMainType;
(function (TitleMainType) {
    TitleMainType["Movie"] = "movie";
    TitleMainType["Series"] = "series";
    TitleMainType["SeriesEpisode"] = "seriesEpisode";
    TitleMainType["TVSpecial"] = "tvSpecial";
    TitleMainType["TVShort"] = "tvShort";
    TitleMainType["TVMovie"] = "tvMovie";
    TitleMainType["Video"] = "video";
})(TitleMainType = exports.TitleMainType || (exports.TitleMainType = {}));
var Language;
(function (Language) {
    Language["Persian"] = "persian";
    Language["English"] = "english";
})(Language = exports.Language || (exports.Language = {}));
var AwardOutcome;
(function (AwardOutcome) {
    AwardOutcome["Winner"] = "winner";
    AwardOutcome["Nominee"] = "nominee";
})(AwardOutcome = exports.AwardOutcome || (exports.AwardOutcome = {}));
var ImageType;
(function (ImageType) {
    ImageType["Poster"] = "poster";
    ImageType["StillFrame"] = "stillFrame";
    ImageType["Event"] = "event";
    ImageType["VideoImage"] = "videoImage";
    ImageType["ProfileImage"] = "profileImage";
    ImageType["Other"] = "other";
})(ImageType = exports.ImageType || (exports.ImageType = {}));
var IMDBPathType;
(function (IMDBPathType) {
    IMDBPathType["Title"] = "title";
    IMDBPathType["Name"] = "name";
})(IMDBPathType = exports.IMDBPathType || (exports.IMDBPathType = {}));
