// movier/dist/src/externalInterfaces/IMDBTitleNextDataInterface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CanRateTypename;
(function (CanRateTypename) {
    CanRateTypename["CanRate"] = "CanRate";
})(CanRateTypename || (CanRateTypename = {}));
var FluffyTypename;
(function (FluffyTypename) {
    FluffyTypename["Name"] = "Name";
})(FluffyTypename || (FluffyTypename = {}));
var PurpleTypename;
(function (PurpleTypename) {
    PurpleTypename["CreditEdge"] = "CreditEdge";
})(PurpleTypename || (PurpleTypename = {}));
var Rating;
(function (Rating) {
    Rating["PG"] = "PG";
    Rating["PG13"] = "PG-13";
    Rating["R"] = "R";
})(Rating || (Rating = {}));
var CertificateTypename;
(function (CertificateTypename) {
    CertificateTypename["Certificate"] = "Certificate";
})(CertificateTypename || (CertificateTypename = {}));
var PlotTextTypename;
(function (PlotTextTypename) {
    PlotTextTypename["Markdown"] = "Markdown";
})(PlotTextTypename || (PlotTextTypename = {}));
var GenresTypename;
(function (GenresTypename) {
    GenresTypename["Genres"] = "Genres";
})(GenresTypename || (GenresTypename = {}));
var PrimaryImageTypename;
(function (PrimaryImageTypename) {
    PrimaryImageTypename["Image"] = "Image";
    PrimaryImageTypename["Thumbnail"] = "Thumbnail";
})(PrimaryImageTypename || (PrimaryImageTypename = {}));
var DisplayNameTypename;
(function (DisplayNameTypename) {
    DisplayNameTypename["LocalizedString"] = "LocalizedString";
})(DisplayNameTypename || (DisplayNameTypename = {}));
var ContentTypeTypename;
(function (ContentTypeTypename) {
    ContentTypeTypename["VideoContentType"] = "VideoContentType";
})(ContentTypeTypename || (ContentTypeTypename = {}));
var IndigoTypename;
(function (IndigoTypename) {
    IndigoTypename["VideoRuntime"] = "VideoRuntime";
})(IndigoTypename || (IndigoTypename = {}));
var StickyTypename;
(function (StickyTypename) {
    StickyTypename["Video"] = "Video";
})(StickyTypename || (StickyTypename = {}));
var TentacledTypename;
(function (TentacledTypename) {
    TentacledTypename["VideoEdge"] = "VideoEdge";
})(TentacledTypename || (TentacledTypename = {}));
var ProductionStatusHistoryTypename;
(function (ProductionStatusHistoryTypename) {
    ProductionStatusHistoryTypename["ProductionStatusHistory"] = "ProductionStatusHistory";
})(ProductionStatusHistoryTypename || (ProductionStatusHistoryTypename = {}));
var RatingsSummaryTypename;
(function (RatingsSummaryTypename) {
    RatingsSummaryTypename["RatingsSummary"] = "RatingsSummary";
})(RatingsSummaryTypename || (RatingsSummaryTypename = {}));
var ReleaseYearTypename;
(function (ReleaseYearTypename) {
    ReleaseYearTypename["YearRange"] = "YearRange";
})(ReleaseYearTypename || (ReleaseYearTypename = {}));
var IndecentTypename;
(function (IndecentTypename) {
    IndecentTypename["Runtime"] = "Runtime";
})(IndecentTypename || (IndecentTypename = {}));
var AboveTheFoldDataTypename;
(function (AboveTheFoldDataTypename) {
    AboveTheFoldDataTypename["Title"] = "Title";
})(AboveTheFoldDataTypename || (AboveTheFoldDataTypename = {}));
var CharacterTypename;
(function (CharacterTypename) {
    CharacterTypename["Character"] = "Character";
})(CharacterTypename || (CharacterTypename = {}));
var EpisodeCreditsTypename;
(function (EpisodeCreditsTypename) {
    EpisodeCreditsTypename["EpisodeCastConnection"] = "EpisodeCastConnection";
})(EpisodeCreditsTypename || (EpisodeCreditsTypename = {}));
var CreditTypename;
(function (CreditTypename) {
    CreditTypename["Cast"] = "Cast";
})(CreditTypename || (CreditTypename = {}));
var PrimaryWatchOptionTypename;
(function (PrimaryWatchOptionTypename) {
    PrimaryWatchOptionTypename["PrimaryWatchOption"] = "PrimaryWatchOption";
})(PrimaryWatchOptionTypename || (PrimaryWatchOptionTypename = {}));
var HilariousTypename;
(function (HilariousTypename) {
    HilariousTypename["MoreLikeThisEdge"] = "MoreLikeThisEdge";
})(HilariousTypename || (HilariousTypename = {}));
var AmbitiousTypename;
(function (AmbitiousTypename) {
    AmbitiousTypename["ImageEdge"] = "ImageEdge";
})(AmbitiousTypename || (AmbitiousTypename = {}));
